File: Python/gamex/store_blizzard.py
import os, platform
from .Blizzard_pb2 import Database

def getPath() -> None:
    system = platform.system()
    if system == 'Windows':
        # windows paths
        home = os.getenv('ALLUSERSPROFILE')
        paths = [os.path.join(home, 'Battle.net', 'Agent')]
    elif system == 'Linux':
        # linux paths
        home = os.path.expanduser('~')
        search = ['.steam', '.steam/steam', '.steam/root', '.local/share/Steam']
        paths = [os.path.join(home, path, 'appcache') for path in search]
    elif system == 'Darwin':
        # mac paths
        home = '/Users/Shared'
        search = ['Battle.net/Agent']
        paths = [os.path.join(home, path, 'data') for path in search]
    else: raise Exception(f'Unknown platform: {system}')
    return next(iter(x for x in paths if os.path.isdir(x)), None)

# get blizzardPaths
blizzardPaths = {}
root = getPath()
if root and os.path.exists(dbPath := os.path.join(root, 'product.db')):
    # query games
    productDb = Database()
    with open(dbPath, 'rb') as f:
        bytes = f.read()
        productDb.ParseFromString(bytes)
        #try: database.ParseFromString(bytes)
        #except InvalidProtocolBufferException: return None
        for app in productDb.ProductInstall:
            # add appPath if exists
            appPath = app.Settings.InstallPath
            if os.path.isdir(appPath): blizzardPaths[app.Uid] = appPath

# print(blizzardPaths)

File: dotnet/Base/GameX.Base/Store_Blizzard.cs
using Google.Protobuf;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Runtime.InteropServices;

namespace GameX.Store
{
    /// <summary>
    /// Store_Blizzard
    /// </summary>
    internal static class Store_Blizzard
    {
        internal static Dictionary<string, string> BlizzardPaths = new Dictionary<string, string>();

        static Store_Blizzard()
        {
            // get dbPath
            var root = GetPath();
            if (root == null) return;
            var dbPath = Path.Combine(root, "product.db");
            if (!File.Exists(dbPath)) return;

            // query games
            Database productDb;
            using var s = File.OpenRead(dbPath);
            try
            {
                productDb = Database.Parser.ParseFrom(s);
            }
            catch (InvalidProtocolBufferException)
            {
                productDb = new Database { ProductInstall = { ProductInstall.Parser.ParseFrom(s) } };
            }
            foreach (var app in productDb.ProductInstall)
            {
                // add appPath if exists
                var appPath = app.Settings.InstallPath;
                if (Directory.Exists(appPath)) BlizzardPaths.Add(app.Uid, appPath);
            }
        }

        static string GetPath()
        {
            IEnumerable<string> paths;
            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
            {
                // windows paths
                var home = Environment.GetFolderPath(Environment.SpecialFolder.CommonApplicationData);
                paths = new[] { Path.Combine(home, "Battle.net", "Agent") };
            }
            else if (RuntimeInformation.OSDescription.StartsWith("android-")) return null;
            else if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
            {
                // linux paths
                var home = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
                var search = new[] { ".steam", ".steam/steam", ".steam/root", ".local/share/Steam" };
                paths = search.Select(path => Path.Join(home, path, "appcache"));
            }
            else if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
            {
                // mac paths
                var home = "/Users/Shared";
                var search = new[] { "Battle.net/Agent" };
                paths = search.Select(path => Path.Join(home, path, "data"));
            }
            else throw new PlatformNotSupportedException();
            return paths.FirstOrDefault(Directory.Exists);
        }
    }
}

